### PR TITLE
Fixes for android and llvm

### PR DESCRIPTION
**Pull request by jiangenj**
- in android file needs to be aboslute path
- add EXCLUDE_PATHS var to bypass multiple defination error
- use readelf instead of objdump to get section offset and size

### CODE REVIEW

**Comment by ianamason** [comment]
What’s with the copyright lines?

**Comment by jiangenj** [comment]
> What’s with the copyright lines?

I'm from quicinc, that's what the company asked for.
I'm fine to remove it if you insist to not have any copyright, but I think license is more important here.

**Comment by ianamason** [comment]
I'm not keen on the copyright.
It would be nice if you could summarize the changes. 
Is that a new environment variable I see? That will need some explanation in the README to be of any use.
Do these changes effect other platforms other than android?

**Comment by jiangenj** [comment]
> I'm not keen on the copyright. It would be nice if you could summarize the changes. Is that a new environment variable I see? That will need some explanation in the README to be of any use. Do these changes effect other platforms other than android?

- updated the readme for the var.
- updated the summary of the pr.
- and there is no side effect as readelf has consistant output with gnu or llvm ones while objdump diffs.

File: wllvm/extraction.py
Comment (ianamason) [inline wllvm/extraction.py:1]: This breaks the cross compilation feature. Read the README more carefully.

File: wllvm/filetype.py
Comment (ianamason) [inline wllvm/filetype.py:1]: Why do we want to hard code /usr/bin as the location of file? What if its in /usr/local/bin?

File: wllvm/extraction.py
Comment (jiangenj) [reply wllvm/extraction.py:1]: No, it won't break it, 
```
cmd_hypcopy = $(OBJCOPY) --prefix-symbols=__kvm_nvhe_ $< $@
```
is from linux kernel, extract-bc doesn't support the --prefix-symbols, so we have to use EXCLUDE_PATHS to exclude some bc from llvm-link.

File: wllvm/filetype.py [hunk 1]
```diff
@@ -34,7 +35,7 @@ def getFileType(cls, fileName):
         Maybe we should use python-magic instead?
         """
         retval = None
-        fileP = Popen(['file', os.path.realpath(fileName)], stdout=PIPE)
+        fileP = Popen(['/usr/bin/file', os.path.realpath(fileName)], stdout=PIPE)
```
Comment (jiangenj) [inline wllvm/filetype.py:38]: so in Android build system, host tool can not be used as PATH var is reset. let me add a check here for /usr/bin/file or /usr/local/bin/file

File: wllvm/filetype.py
Comment (jiangenj) [reply wllvm/filetype.py:1]: use abs path first

**Comment by jiangenj** [comment]
can u review again?

File: wllvm/filetype.py (hunk 1 above)
Comment (ianamason) [reply wllvm/filetype.py]: So I'm not going to merge in hard coded paths like this. If you look at the code base you will notice that most tools we use (via Popen) have the feature that you can specify the full path via an environment variable.  If it was well documented I would include this approach. But not a hardcoded path hack.

File: wllvm/extraction.py
Comment (ianamason) [inline wllvm/extraction.py:1]: Making readelf the default will break other platforms. You need to think about preserving behavior not just changing behavior. Again I am not going to merge in changes that are hacky or break support of other platforms.

I'm not sure that the  EXCLUDE_PATHS hack is needed. Why not just extract the bitcode out as a module and then do the linking yourself by hand?

**Comment by ianamason** [comment]
I added another comment about the file type, but it seems to have disappeared, so here it is again.

I'm not going to merge in hard coded paths. When we use a tool (via Popen) we often provide the ability to include its location via a environment variable. This approach would work for 'file' but it would need to be documented. Something like WLLVM_FILE_TOOL_PATH

**Pull request closed without merging**